prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1333 @@
+				     util_cache_entry_flag_t flag)
+{
+	return ((entry->state & flag) != 0);
+}
+
+static inline void util_entry_set_flag(struct util_mr_cache_entry *entry,
+				    util_cache_entry_flag_t flag)
+{
+	entry->state = entry->state | flag;
+}
+
+static inline int util_entry_is_retired(struct util_mr_cache_entry *entry)
+{
+	return util_entry_is_flag(entry, UTIL_CE_RETIRED);
+}
+
+static inline int util_entry_is_merged(struct util_mr_cache_entry *entry)
+{
+	return util_entry_is_flag(entry, UTIL_CE_MERGED);
+}
+
+static inline int util_entry_is_unmapped(struct util_mr_cache_entry *entry)
+{
+	return util_entry_is_flag(entry, UTIL_CE_UNMAPPED);
+}
+
+static inline void util_entry_set_retired(struct util_mr_cache_entry *entry)
+{
+	util_entry_set_flag(entry, UTIL_CE_RETIRED);
+}
+
+static inline void util_entry_set_merged(struct util_mr_cache_entry *entry)
+{
+	util_entry_set_flag(entry, UTIL_CE_MERGED);
+}
+
+static inline void util_entry_set_unmapped(struct util_mr_cache_entry *entry)
+{
+	util_entry_set_flag(entry, UTIL_CE_UNMAPPED);
+}
+
+/**
+ * Key comparison function for finding overlapping memory
+ * registration cache entries
+ *
+ * @param[in] x key to be inserted or found
+ * @param[in] y key to be compared against
+ *
+ * @return    -1 if it should be positioned at the left, 0 if it overlaps,
+ *             1 otherwise
+ */
+static int util_find_overlapping_addr(void *x, void *y)
+{
+	struct util_mr_cache_key *to_find  = (struct util_mr_cache_key *)x;
+	struct util_mr_cache_key *to_compare = (struct util_mr_cache_key *)y;
+	uint64_t to_find_end = to_find->address + to_find->length;
+	uint64_t to_compare_end = to_compare->address + to_compare->length;
+
+	/* format: (x_addr,  x_len) - (y_addr,  y_len) truth_value
+	 *
+	 * case 1: (0x1000, 0x1000) - (0x1400, 0x0800) true
+	 * case 2: (0x1000, 0x1000) - (0x0C00, 0x0800) true
+	 * case 3: (0x1000, 0x1000) - (0x1C00, 0x0800) true
+	 * case 4: (0x1000, 0x1000) - (0x0C00, 0x2000) true
+	 * case 5: (0x1000, 0x1000) - (0x0400, 0x0400) false
+	 * case 6: (0x1000, 0x1000) - (0x2400, 0x0400) false
+	 */
+	if (!((to_find_end < to_compare->address) ||
+	      (to_compare_end < to_find->address)))
+		return 0;
+
+	/* left */
+	if (to_find->address < to_compare->address)
+		return -1;
+
+	return 1;
+}
+
+/**
+ * Key comparison function for memory registration caches
+ *
+ * @param[in] x key to be inserted or found
+ * @param[in] y key to be compared against
+ *
+ * @return    -1 if it should be positioned at the left, 0 if the same,
+ *             1 otherwise
+ */
+static inline int util_mr_cache_key_comp(void *x, void *y)
+{
+	struct util_mr_cache_key *to_insert  = (struct util_mr_cache_key *)x;
+	struct util_mr_cache_key *to_compare = (struct util_mr_cache_key *)y;
+
+	if (to_compare->address == to_insert->address)
+		return 0;
+
+	/* to the left */
+	if (to_insert->address < to_compare->address)
+		return -1;
+
+	/* to the right */
+	return 1;
+}
```

Comment:
Those functions are different. One attempts to find the exact memory registration, while the other attempts to identify regions for which it could overlap with the proposed region.